src/Type/Constant/ConstantArrayType.php

```diff
@@ -839,14 +839,29 @@ public function popArray(): Type
 
 	private function reverseConstantArray(TrinaryLogic $preserveKeys): self
 	{
-		$keyTypesReversed = array_reverse($this->keyTypes, true);
-		$keyTypes = array_values($keyTypesReversed);
-		$keyTypesReversedKeys = array_keys($keyTypesReversed);
-		$optionalKeys = array_map(static fn (int $optionalKey): int => $keyTypesReversedKeys[$optionalKey], $this->optionalKeys);
+		if ($preserveKeys->yes()) {
+			$keyTypesReversed = array_reverse($this->keyTypes, true);
+			$keyTypes = array_values($keyTypesReversed);
+			$keyTypesReversedKeys = array_keys($keyTypesReversed);
+			$optionalKeys = array_map(static fn (int $optionalKey): int => $keyTypesReversedKeys[$optionalKey], $this->optionalKeys);
 
-		$reversed = new self($keyTypes, array_reverse($this->valueTypes), $this->nextAutoIndexes, $optionalKeys, TrinaryLogic::createNo());
+			return new self($keyTypes, array_reverse($this->valueTypes), $this->nextAutoIndexes, $optionalKeys, TrinaryLogic::createNo());
+		}
+
+		$builder = ConstantArrayTypeBuilder::createEmpty();
+		for ($i = count($this->keyTypes) - 1; $i >= 0; $i--) {
+			$offsetType = $this->keyTypes[$i]->isInteger()->no()
+				? $this->keyTypes[$i]
+				: null;
+			$builder->setOffsetValueType($offsetType, $this->valueTypes[$i], $this->isOptionalKey($i));
+		}
 
-		return $preserveKeys->yes() ? $reversed : $reversed->reindex();
+		$array = $builder->getArray();
+		if (!$array instanceof self) {
+			throw new ShouldNotHappenException();
+		}
+
+		return $array;
```

Comment:
it would be possible to use the same implementation with the builder also  if $preserveKeys->yes(), but the current implementation feels might be slightly more optimised I suppose? not sure.

Reply:
I remove the old logic, it feels like it makes things unnecessarily more complex and tbh I don't even know how much more optimised it is